File: articles/statistics-uncertainty-measurement-biology/python/calibration_curve.py
"""
Calibration curve workflow.

Run:
    python python/calibration_curve.py
"""

from pathlib import Path

import pandas as pd

from statistics_measurement_core import concentration_from_response, fit_linear_calibration


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "calibration_standards.csv"


def main() -> None:
    standards = pd.read_csv(DATA_PATH)

    fit = fit_linear_calibration(standards["concentration"], standards["response"])

    unknown_response = 6.25
    estimated_concentration = concentration_from_response(unknown_response, fit)

    summary = pd.DataFrame(
        {
            "intercept": [fit.intercept],
            "slope": [fit.slope],
            "r_squared": [fit.r_squared],
            "unknown_response": [unknown_response],
            "estimated_concentration": [estimated_concentration],
        }
    )

    print(summary.round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/statistics-uncertainty-measurement-biology/r/calibration_curve.R
# Calibration curve workflow in R.

data_path <- file.path("data", "calibration_standards.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "calibration_standards.csv")
}

standards <- read.csv(data_path)

fit <- lm(response ~ concentration, data = standards)

unknown_response <- 6.25

estimated_concentration <- (
  unknown_response - coef(fit)[["(Intercept)"]]
) / coef(fit)[["concentration"]]

summary_df <- data.frame(
  intercept = coef(fit)[["(Intercept)"]],
  slope = coef(fit)[["concentration"]],
  r_squared = summary(fit)$r.squared,
  unknown_response = unknown_response,
  estimated_concentration = estimated_concentration
)

print(round(summary_df, 5))
